File: projects/grind-75/python/string_lc_questions/valid_parenthesis.py
class Solution:
    @classmethod
    def isValid(self, s):
        hmap = {")": "(", "}": "{", "]": "["}
        stack = []
        for c in s:
            if c in hmap:
                if not stack or stack.pop() != hmap[c]:
                    return False
            else:
                stack.append(c)
        return not stack

File: projects/grind-75/typescript/string/valid-parenthesis.ts
export function isValid(s: string): boolean {
  const stack: string[] = [];
  const hmap: { [key: string]: string } = { ']': '[', '}': '{', ')': '(' };

  for (const c of s) {
    if (Object.values(hmap).includes(c)) {
      stack.push(c);
    } else if (Object.keys(hmap).includes(c)) {
      if (!stack || hmap[c] !== stack.pop()) {
        return false;
      }
    } else {
      return false;
    }
  }
  return stack.length === 0;
}
